File: cplusplus/Bridge_NullLogger.cpp
/// @file
/// @brief
/// Implementation of the @ref DesignPatternExamples_cpp::NullLogger "NullLogger"
/// class factory used in the @ref bridge_pattern.

#include "Bridge_NullLogger.h"

namespace // Anonymous
{
    /// <summary>
    /// Represents a logger that throws away anything sent its way.
    /// </summary>
    class NullLoggerImpl : public DesignPatternExamples_cpp::ILogger
    {
    public:
        /// <summary>
        /// Constructor.
        /// </summary>
        NullLoggerImpl()
        {
        }

        // ILogger Members

        void LogTrace(const std::string&)
        {
        }

        void LogInfo(const std::string&)
        {
        }

        void LogError(const std::string&)
        {
        }
    };

} // end anonymous namespace


namespace DesignPatternExamples_cpp
{
    
    std::unique_ptr<ILogger> NullLogger::CreateLogger()
    {
        return std::make_unique<NullLoggerImpl>();
    }
    
} // end namespace

File: csharp/Bridge_NullLogger.cs
/// @file
/// @brief
/// The @ref DesignPatternExamples_csharp.NullLogger "NullLogger"
/// class as used in the @ref bridge_pattern "Bridge pattern".

using System;

namespace DesignPatternExamples_csharp
{
    /// <summary>
    /// Represents a logger that throws away anything sent its way.
    /// </summary>
    internal class NullLogger : ILogger, IDisposable
    {
        /// <summary>
        /// Constructor.
        /// </summary>
        private NullLogger()
        {
        }


        #region ILogger Members

        void ILogger.LogTrace(string msg)
        {
        }

        void ILogger.LogInfo(string msg)
        {
        }

        void ILogger.LogError(string msg)
        {
        }

        #endregion

        #region IDisposable Members

        /// <summary>
        /// Nothing to do for a graceful shutdown.
        /// </summary>
        void IDisposable.Dispose()
        {
        }

        #endregion


        /// <summary>
        /// Create an instance of a null logger, a logger that doesn't do anything.
        /// </summary>
        /// <returns>An instance of an ILogger object.</returns>
        public static ILogger CreateNullLogger()
        {
            return new NullLogger();
        }
    }
}
